### DISCUSSION TITLE
Would it make sense for the reducer closure to be marked as @MainActor?

### DISCUSSION THREAD

**Discussion by danieleformichelli** [pointfreeco/swift-composable-architecture#1272] [General]
Hello 👋 

Given the reducer usually runs on the main thread, would it make sense to make [the reducer closure](https://github.com/pointfreeco/swift-composable-architecture/blob/d41f36c707206625ca310ff4f929e3a350445733/Sources/ComposableArchitecture/Reducer.swift#L56) `@MainActor`, or at least have the possibility to chose whether it needs to be a `@MainActor` or not?

Right now you can't access other non-async `@MainActor`s directly in the reducer code because you would need to `await` for it, but if the reducer itself would be a `@MainActor` it wouldn't need to await to access another `@MainActor`, right?

Thank you 🙏

**Comment by mbrandonw** [comment]
Hi @danyf90, it is true that currently reducers are only meant to be run on the main thread, but marking it explicitly as `@MainActor` would be a breaking change for everyone using the library. We are moving forward with concurrency changes slowly so that we can minimize breakage when updating to the newest version of the library. For example, all of the new concurrency tools we introduced in 0.39 are fully backwards compatible and did not require any changes to existing TCA applications.

So, someday in the future we could make reducers `@MainActor`. But, what we'd really like to do is provide an even deeper integration of the library with Swift concurrency. This may mean turning `Store` into an actor or non-main global actor, and making `ViewStore` into a `@MainActor`. This would be a huge breaking change, and so we need to time it correctly and roll it out slowly, but once we achieve that then reducers do _not_ need to be made `@MainActor`, which would be nice.

As an aside, I'm curious what APIs you are using in the reducer that are marked as `@MainActor`. It's possible that those APIs should be interacted with in an effect anyway so that they are controllable during tests and previews, in which case you are provided an async context to `await`, such as in `Effect.task`/`.run`/`.fireAndForget`.

**Reply by danieleformichelli** [reply]
Thanks for the prompt reply @mbrandonw! 🙏 

I don't have a real example, but I was thinking of some external library/API exposing some state or values through `@MainActor`, for which it would be handy to access the value without the need to use an effect.

> It's possible that those APIs should be interacted with in an effect anyway so that they are controllable during tests and previews, in which case you are provided an async context to await

If both the reducer and the dependency are marked as `@MainActor`, there is basically no async context involved so the reducer would be testable, isn't it?

Can't wait to see how TCA will further evolve around the new concurrency model 🚀

**Reply by fatbobman** [reply]
Hi, @mbrandonw . I currently include the NSManagedObject obtained from the viewContext in the State and use a computed property to get its contents when needed. Since the current Reducer is running on the main thread, everything works fine.
If TCA creates a GlobalActor to annotate the Reducer in the future, can it have the ability to still set it as the MainActor?

**Reply by Alex293** [reply]
It would be better to listen for db changes and publish them instead of having references in your state. You can then have plain swift struct in your state. To get back to the subject this make you in control of which actor /thread you listen to the changes and TCA could use its own actor. I think if TCA migrate toward using actor the store will want to have exclusivity on its actor to avoid re-entrancy thought as UIKit and SwiftUI rely on MainActor store output will stay on MainActor.

**Reply by fatbobman** [reply]
@Alex293 Yes, that's exactly what I'm worried about. I shouldn't assume that in the future the Reducer will necessarily run on the main thread. But since ViewStore can be guaranteed to run on the main thread, maybe I can perform operations in ViewState that require the participation of MainActor.

**Reply by Alex293** [reply]
Either way I wouldn't store NSManagedObject directly in state and listen and map those to structs.
I'm pretty your views won't update if only NSManaged objects are updated.

**Reply by fatbobman** [reply]
<img width="988" alt="image" src="https://user-images.githubusercontent.com/55673881/201919478-79236301-b51f-4e2c-955c-5dff0c5ddc94.png">
For efficiency reasons, I can't convert all query results into struct and put them into TCA's State. I only keep a reference to them in State. Each object is passed to a child Reducer via ForEachStore . And trigger the update state through a seed.
It's working fine so far. But considering that Reducer may not be able to run on non-main thread in the future (my NSMangedObjects are obtained from viewContext), I am going to fire fault this process in viewState.
There is currently no problem passing them in different threads as long as the data in the NSManagedObject properties is not read.

**Reply by gerfarfel** [reply]
I think that Core Data really is a good example of a use case where a guarantee of having the reducer run on the main thread (or at least on a known specified thread) could be very helpful. In the app I'm working on, I'd like to be able to update TCA state with data from my Core Data store synchronously in the reducer, with no thread hops. One reason for this is that it would make coordination with my undo manager (implemented via a higher-order reducer) much simpler, reducing the likelihood of bugs. (And btw, I don't store Core Data managed objects in my TCA state, so that's not the issue.) @mbrandonw, your reply to the original question implies that we shouldn't assume that we will be able to specify the reducer's thread in future versions of TCA. I'm finding that working with that assumption is making my code more complex (and probably more error-prone). I'm pretty sure that I would mark my reducers as `@MainActor` if I was confident that I would be able to continue to do that in future versions of TCA – but as things stand, I'm not.

**Reply by mbrandonw** [reply]
Hi @gerfarfel, it seems too early to make any strong guarantees of when/where reducers will run, especially with respect to Swift's concurrency tools. After all, there still isn't an officially accepted proposal for custom executors, which is pretty much a requirement for reasonably interacting with core data in an async context.

We are keeping our eye on all the various proposals to see how things are progressing, but we cannot make reducers or stores `@MainActor` right now because it would be a hugely breaking change. At the very least we _do_ guarantee that stores are single threaded, so as long as you send all actions on the main thread then all reducers will also run on the main thread.
